elkscmd/gui/app.c

```diff
@@ -61,132 +59,81 @@ void A_InitPalette(void)
-    paletteButtons[9].OnClick = G_SetBushSize;
-    paletteButtons[9].data1 = 12;
-    paletteButtons[9].render = true;
-
-    // Save
-    paletteButtons[10].name = "SaveButton";
-    paletteButtons[10].box.x = SCREEN_WIDTH + 10;
-    paletteButtons[10].box.y = 300;
-    paletteButtons[10].box.w = 32;
-    paletteButtons[10].box.h = 32;
-    paletteButtons[10].OnClick = G_SaveButtonOnClick;
-    paletteButtons[10].render = true;
-    paletteButtons[10].fileName = LIBPATH "save.bmp";
-
-    // Clear Screen
-    paletteButtons[11].name = "Cls";
-    paletteButtons[11].box.x = SCREEN_WIDTH + 10 + 32*2;
-    paletteButtons[11].box.y = 300;
-    paletteButtons[11].box.w = 32;
-    paletteButtons[11].box.h = 32;
-    paletteButtons[11].OnClick = G_ClearScreen;
-    paletteButtons[11].render = true;
-    paletteButtons[11].fileName = LIBPATH "cls.bmp";
-
-    // Quit
-    paletteButtons[12].name = "QuitButton";
-    paletteButtons[12].box.x = SCREEN_WIDTH + 10 + 32*3;
-    paletteButtons[12].box.y = 300;
-    paletteButtons[12].box.w = 32;
-    paletteButtons[12].box.h = 32;
-    paletteButtons[12].OnClick = G_QuitButtonOnClick;
-    paletteButtons[12].render = true;
-    paletteButtons[12].fileName = LIBPATH "quit.bmp";
+    // --- Helper macro to initialize buttons ---
+#define INIT_BUTTON(ID, LABEL, X, Y, W, H, CLICK, RENDER) \
+    paletteButtons[ID].name = LABEL; \
+    paletteButtons[ID].box.x = (X); \
+    paletteButtons[ID].box.y = (Y); \
+    paletteButtons[ID].box.w = (W); \
+    paletteButtons[ID].box.h = (H); \
+    paletteButtons[ID].OnClick = CLICK; \
+    paletteButtons[ID].render = RENDER;
+
+    #if UNUSED
+        INIT_BUTTON(0, "BrightnessSelector", 775, 10, 16, 64, G_BrightnessButtonOnClick, 0);
+    #endif
+
+    // --- Color Picker ---
+    INIT_BUTTON(1, "ColorPicker", CANVAS_WIDTH + 10, 10, 128, 128, G_ColorPickerOnClick, 0);
+
+    // --- Brush Size Buttons (8 total in two rows) ---
+    {
+        int brushSizes[8] = {1, 2, 3, 4,
+                                   6, 8, 10, 12};
+        char* brushNames[8] = {
+            "BushSize1", "BushSize2", "BushSize3", "BushSize4",
+            "BushSize5", "BushSize6", "BushSize7", "BushSize8"
+        };
+
+        for (int i = 0; i < 8; ++i) {
+            int index = i + 2;
+            int col = i % 4;
+            int row = i / 4;
+            INIT_BUTTON(index,
+                        brushNames[i],
+                        CANVAS_WIDTH + 11 + 32 * col,
+                        195 + row * 32,
+                        30, 30,
+                        G_SetBushSize, 1);
+            paletteButtons[index].data1 = brushSizes[i];
+        }
+    }
+
+    // --- Utility Buttons ---
+    {
+        int utilCount = 5;
+        int indices[]   = {10, 11, 12, 13, 14};
+        char* names[]   = {"SaveButton", "FillButton", "BrushButton", "QuitButton", "Cls"};
+        int offsetX[]   = {0,  32,   0, 32 * 3, 32 * 3};
+        int offsetY[]   = {0, -32, -32,      0,    -32};
```

Comment:
Are you aware how expensive this is, to declare auto initialized variables? The initializations have to be saved in the data segment, but then copied via an internal call to memcpy to get them on the stack; IMO lots of extra code for nothing.

I recommend adding 'static' to the declarations which achieves the same result with lots less code. Same for the variables on lines 81-85.

(You can use `ia16-elf-objdump -D -r -Mi8086 app.oaj` to see code generated).

Reply:
Oh, no I was not aware of this. I was just looking for a way to simplify new buttons addition but I couldn’t achieve it with macros only. 
I will rewrite it on next PR

Reply:
I like the new coding and ease of inspection of the initializers for each case, I don't think any rewrite is necessary except just applying static prefix to the declarations of the arrays (only).